[close #33907] Error when using "recyclable" cache keys with a store that does not support it

schneems opened the pull request:
If you are using the "in cache versioning" also known as "recyclable cache keys" the cache store must be aware of this scheme, otherwise you will generate cache entries that never invalidate.

This PR adds a check to the initialization process to ensure that if recyclable cache keys are being used via 

```
config.active_record.cache_versioning = true
```

Then the cache store needs to show that it supports this versioning scheme. Cache stores can let Rails know that they support this scheme by adding a method `supports_in_cache_versioning?` and returning true.

schneems committed:
[close #33907] Error when using "recyclable" cache keys with a store that does not support it

If you are using the "in cache versioning" also known as "recyclable cache keys" the cache store must be aware of this scheme, otherwise you will generate cache entries that never invalidate.

This PR adds a check to the initialization process to ensure that if recyclable cache keys are being used via 

```
config.active_record.cache_versioning = true
```

Then the cache store needs to show that it supports this versioning scheme. Cache stores can let Rails know that they support this scheme by adding a method `supports_in_cache_versioning?` and returning true.

schneems commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -88,6 +88,29 @@ class Railtie < Rails::Railtie # :nodoc:
+    initializer "Check for cache versioning support" do
+      config.after_initialize do |app|
+        ActiveSupport.on_load(:active_record) do
+          if app.config.active_record.cache_versioning && defined?(Rails) && Rails.cache
```
good point, it's in a..railtie. Fixed.

jeremy commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -88,6 +88,29 @@ class Railtie < Rails::Railtie # :nodoc:
+
+You're using a cache store `#{Rails.cache.class}` that does not support
+"recyclable" cache keys, also known as "in cache versioning". To
+fix this issue either disable "recyclable" cache keys by setting:
```
Let's stick with cache versioning terminology as a conceptual anchor and rely on the guides to discuss in detail. We can drop quoted terms. Let's be clear about what we're asking of the cache store, too, e.g.

> You're using a cache store that doesn't support native cache versioning. Your best option is to upgrade to a newer version of `#{Rails.cache.class}` that supports cache versioning (`#{Rails.cache.class}.supports_cache_versioning? #=> true`). Next best, switch to a different cache store that does support cache versioning: https://guides.rubyonrails.org/caching_with_rails.html#cache-stores. To keep using the current cache store, you can turn off cache versioning entirely: `config.active_record.cache_versioning = false`.

jeremy commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -88,6 +88,29 @@ class Railtie < Rails::Railtie # :nodoc:
+      config.after_initialize do |app|
+        ActiveSupport.on_load(:active_record) do
+          if app.config.active_record.cache_versioning && Rails.cache
+            unless Rails.cache.try(:supports_in_cache_versioning?)
```
Let's use a class method and normalize naming: `Rails.cache.class.supports_cache_versioning?`

schneems replied:
I made a comment about the naming in basecamp. I think it's confusing. All caches are versioned already via cache key, the conceptual change is that the versioning is now __in__ the cache which is what i'm trying to convey here with that word. I can standardize on this though.

jeremy commented on activesupport/lib/active_support/cache/file_store.rb:
```diff
@@ -26,6 +26,13 @@ def initialize(cache_path, options = nil)
+      def supports_in_cache_versioning?
+        true
+      end
+
```
Let's do a `supports_cache_versioning?` class method for these.

Comment may be unwrapped to `# Advertise cache versioning support.`

(Worth noting that this duck typing won't cover subclasses of these cache stores.)

jeremy commented on railties/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+*   Raise an error when "recyclable cache keys" are being used by a cache store
+    that does not explicitly support it.
```
Let's describe what cache stores need to do (implement `supports_cache_versioning` class method) to advertise support.

schneems replied:
For this what are you thinking? Maybe a section in the guide?

schneems committed:
Switch to supports_cache_versioning? check to a class method

- Moving the `supports_cache_versioning?` check to a class method. 
- Shorten the method doc. 
- Expand on the error message.
